bigtable_input.py

```diff
@@ -0,0 +1,91 @@
+"""Read Minigo game examples from a Bigtable.
+"""
+
+from google.cloud import bigtable
+from google.cloud.bigtable import row_filters
+import logging
+import os
+import struct
+import tensorflow as tf
+
+
+_project_name = os.environ['PROJECT']
+_instance_name = os.environ['CBT_INSTANCE']
+_table_name = os.environ['CBT_TABLE']
+
+_bt_table = bigtable.Client(
+    _project_name).instance(
+        _instance_name).table(
+            _table_name)
+
+_tf_table = tf.contrib.cloud.BigtableClient(
+    _project_name,
+    _instance_name).table(
+        _table_name)
+
+
+_features = {
+    'x': tf.FixedLenFeature((), tf.string, ''),
+    'pi': tf.FixedLenFeature((), tf.string, ''),
+    'outcome': tf.VarLenFeature(dtype=tf.float32)
+    }
+
+
+def get_latest_game_number():
+  """Return the number of the next game to be stored.
+  """
+  table_state = _bt_table.read_row('table_state')
+  value = table_state.cells['metadata'][b'game_counter'][0].value
+  return struct.unpack('>q', value)[0]
+
+
+def get_game_range_row_names(game_begin, game_end):
+  """Get the row range containing the given games.
+
+  Sample row name:
+    g_0000000001_m000001
+
+  To capture the range of all moves in the two given games,
+  the end row will Need to go up to g_00..(N+1).
+  """
+  row_fmt = 'g_{:0>10}_'
+  return row_fmt.format(game_begin), row_fmt.format(game_end + 1)
+
+
+def get_unparsed_moves_from_last_n_games(n, probability,
+                                         column_family='tfexample',
+                                         column='example'):
+  latest_game = int(get_latest_game_number())
+  logging.info("Latest game: %s", latest_game)
+  if latest_game == 0:
+    return 0
+
+  start = int(max(0, latest_game - n))
+  start_row, end_row = get_game_range_row_names(start, latest_game)
+  logging.info("Row range: %s - %s", start_row, end_row)
+  ds = _tf_table.parallel_scan_range(start_row, end_row,
+                                     probability=probability,
+                                     columns=[(column_family, column)])
+  return ds.map(lambda row_name, s: s)
+
+
+def get_moves_from_last_n_games(n, probability,
+                                column_family='tfexample',
+                                column='example'):
+  ds = get_unparsed_moves_from_last_n_games(n, probability,
+                                            column_family, column)
+  return ds.map(tf.parse_single_example)
+
+
+def realize_dataset(ds):
+  iterator = ds.make_initializable_iterator()
```

Comment:
Is this for debug? If so please give it a docstring explaining

Reply:
It definitely is! (See the above problem of counting rows.) I'll document it as such. Also, on reflection, it's better to call it `materialize_dataset`, which I will do.